File: ml-benchmark/synduce/list/line_of_sight.f
Inductive List = single Int | cons {Int, List};
Inductive CList = elt Int | cat {CList, CList};

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: List. (fix (
  \f: List -> {Int, Int, Bool}. \xs: List.
  match xs with
    single a -> {a, a, true}
  | cons {hd, tl} -> 
    let result = f tl in
    {hd, max result.2 hd, > hd result.2}
  end
) xs).3;

cat_list = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    single a -> cons {a, ys}
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    elt a -> single a
  | cat {a, b} -> cat_list (f a) (f b)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/line_of_sight.ml
type tlist = Single of int | Cons of int * tlist
type clist = Elt of int | Cat of clist * clist

let max a b = if a < b then b else a

let spec xs =
  let rec f xs =
    match xs with
    | Single a -> (a, a, true)
    | Cons (hd, tl) ->
        let result = f tl in
        match result with
        | (_, r2, _) -> (hd, max r2 hd, hd > r2)
  in
  match f xs with
  | (_, _, b) -> b

let rec first = function
| Single a -> a
| Cons (a, _) -> a 

let rec last = function
| Single a -> a
| Cons (_, t) -> last t

let rec maximum = function 
| Single a -> a 
| Cons (h, t) -> max h (maximum t)

let rec cat_list xs ys =
  match xs with
  | Single a -> Cons (a, ys)
  | Cons (a, b) -> Cons (a, cat_list b ys)

val repr: clist -> tlist compress
let rec repr xs =
  match xs with
  | Elt a -> Single a
  | Cat (a, b) -> cat_list (repr a) (repr b)

let program x = spec (repr x)